Move mpool from node API to plumbing

anorth opened the pull request:
Closes #2017. A prefactor for #1933

rosalinekarr commented on plumbing/api.go:
```diff
@@ -60,7 +63,7 @@ func New(deps *APIDeps) *API {
 		chain:        deps.Chain,
 		config:       deps.Config,
-		messagePool:  deps.MessagePool,
+		mpool:        mpool.New(deps.MessagePool, deps.Pubsub),
```
Can we call it `msgPool` for consistency?

phritz commented on plumbing/api.go:
```diff
@@ -29,7 +31,7 @@ type API struct {
 	chain        *chn.Reader
 	config       *cfg.Config
-	messagePool  *core.MessagePool
+	mpool        *mpool.Pool
```
mpl is idiomaticer, but up to you

phritz commented on plumbing/mpool/pool.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+// Pool interfaces to the core message pool
+type Pool struct {
```
mpool.pool stutters pretty bad. we could actually stick this in the msg package and have it be msg.pool. alternatively, remove could just be implemented by having the plumbing api call directly into the core.messagepool. then this could be an mpl.lser or mpl.viewer.

anorth replied:
Gone

phritz commented on plumbing/mpool/pool.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+// Pending lists un-mined messages in the pool
+func (ls *Pool) Pending(ctx context.Context, messageCount uint) ([]*types.SignedMessage, error) {
```
should probably say what messagecount does and is for, waits for that many messages and its used for testing. 

actually this is a pretty weird thing to have in the interface. if we didn't have to support it we could just have the core.messagepool in the plumbing and we wouldn't need this wrapper. if pubsub was available in plumbing then a porcelain call could implement this wait functionality by depending on the plumbing pending() and messagepubsub() calls.

anorth replied:
Done!

phritz commented on plumbing/mpool/pool.go:
```diff
@@ -0,0 +1,49 @@
+func (ls *Pool) Pending(ctx context.Context, messageCount uint) ([]*types.SignedMessage, error) {
+	pending := ls.pool.Pending()
+	if len(pending) < int(messageCount) {
+		subscription, err := ls.pubsub.Subscribe(msg.Topic)
```
maybe subscribe before calling pending in case a message arrives in between these calls as written?

anorth replied:
I've added an additional call to pending before the first `Next()`, preserving the fast path that avoids subscribing at all unless necessary.

phritz commented on plumbing/mpool/pool.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+// Pending lists un-mined messages in the pool
+func (ls *Pool) Pending(ctx context.Context, messageCount uint) ([]*types.SignedMessage, error) {
```
blocking: we should have a unit test for the message count functionality

phritz commented:
the more i think about the message counting the more i think we should push it into the tests somehow, and out of the interface...

anorth commented:
Ok, I'm going to pause this PR and move pubsub into plumbing first. #2018

anorth commented:
I have rebased this against #2019 and pulled the wait functionality up to porcelain, where it's now tested. I removed the `plumbing/mpl` package in favour of just putting the `core.MessagePool` on the plumbing struct.

Ready for review again.

phritz commented on porcelain/mpool_test.go:
```diff
@@ -0,0 +1,95 @@
+
+import (
+	"context"
+	"gx/ipfs/QmVRxA4J3UPQpw74dLrQ6NJkfysCA1H4GU28gVpXQt9zMU/go-libp2p-pubsub"
```
this one belongs below

phritz commented on plumbing/ps/testing.go:
```diff
@@ -0,0 +1,83 @@
+		return nil, s.err
+	}
+	select {
+	case msg := <-s.pending:
```
pointing this out now not bc i think it is important here bc it doesnt seem so, but because this is one of the sharp edges of go channels: receiving on a channel actually returns two values. the second is a bool signifying the channel is still open. if the second is false then the channel was closed by caller and the first received value is the zero value for the type. i have been bitten by this many times so make a point of mentioning it to anyone who might not have had the pleasure of hitting it.
